datatables.qmd

```diff
@@ -0,0 +1,255 @@
+Recently it was announced that the 2.0 release of [pandas](https://pandas.pydata.org) will allow for Arrow tables.
+
+Arrow specifies a language-independent tabular memory format that provides many desirable properties, such as provision for missing data and compact representation of categorical data vectors, for data science.
+The memory format also defines a file format for storing and exchanging data tables.
+Because the memory format is essentially the same as the file format, Arrow files can be [memory-mapped](https://en.wikipedia.org/wiki/Memory-mapped_file) providing very fast read speeds.
+
+Furthermore, the Arrow Project provides a reference implementation of the Arrow format and tools for manipulating data in that format as a C++ library, which is used by implementations in several other languages, including C, C#, Go, Java, JavaScript, MATLAB, Python, R, and Ruby.
+The Julia implementation in [Arrow.jl](https://github.com/apache/julia-arrow) does not call functions in the C++ library.
+Instead it implements the format in Julia code in such a way that Arrow vectors behave like native Julia vectors.
+That is, Arrow vectors are a subtype of `AbstractVector`.
+A similar approach is taken in the implementations of the Arrow format for the [Rust](https://rustlang.org) language.
+
+The data sets in the `MixedModels` package and the auxillary data sets used in this book are stored in the Arrow file format and retrieved as `Arrow.Table`s.
+There are many examples throughout this book of loading such data sets.
+
+```{julia}
+contra = MixedModels.dataset(:contra)
+```
+
+Often, for ease of access and for display, we convert the Arrow table to a `Table`, which, contrary to convention, is a type defined in [TypedTables.jl](https://github.com/JuliaData/TypedTables.jl) and not in [Tables.jl](https://github.com/JuliaData/Tables.jl).
+
+Before going into detail about the properties and use of the `Table` type, let us first discuss the role of [Tables.jl](https://github.com/JuliaData/Tables.jl).
+
+## Tables.jl provides an "interface for tables" {#sec-Tablesjl}
+
+An important characteristic of any system for working with data tables is whether the table is stored in memory column-wise or row-wise.
+
+As described above, most implementations of data frames for data science store the data column-wise.
+
+In *relational database management systems* (RDBMS), such as [PostgreSQL](https://postgresql.org), [SQLite](https://sqlite.org), and a multitude of commerical systems, a data table, called a *relation*, is typically stored row-wise.
+Such systems typically use [SQL](https://en.wikipedia.org/wiki/SQL), the *structured query language*, to define and access the data in tables, which is why that acronym appears in many of the names.
+There are exceptions to the row-wise rule, such as [DuckDB](https://duckdb.org), an SQL-based RDBMS, that, as mentioned above, represents relations as Arrow tables.
```

Comment:
Not relevant for the example here, but one of the things that [Snowflake](https://www.snowflake.com/en/) does to achieve performance on data-science workloads is the use of columnar storage.